src/storage/mvcc/consistency_check.rs

```diff
@@ -0,0 +1,462 @@
+    pub fn new<F>(f: F, from: Option<&[u8]>, to: Option<&[u8]>, ob: Ob) -> Result<Self>
+    where
+        F: Fn(&str, IterOptions) -> Result<Iter>,
+    {
+        let from = from.unwrap_or(keys::DATA_MIN_KEY);
+        let to = to.unwrap_or(keys::DATA_MAX_KEY);
+        let key_builder = |key: &[u8]| -> Result<Option<KeyBuilder>> {
+            if !keys::validate_data_key(key) && key != keys::DATA_MAX_KEY {
+                return Err(box_err!("non-mvcc area {}", hex::encode_upper(key)));
+            }
+            Ok(Some(KeyBuilder::from_vec(key.to_vec(), 0, 0)))
+        };
+
+        let iter_opts = IterOptions::new(key_builder(from)?, key_builder(to)?, false);
+        let gen_iter = |cf: &str| -> Result<Iter> {
+            let mut iter = f(cf, iter_opts.clone())?;
+            box_try!(iter.seek(SeekKey::Key(from)));
+            Ok(iter)
+        };
+
+        Ok(MvccInfoScanner {
+            lock_iter: gen_iter(CF_LOCK)?,
+            default_iter: gen_iter(CF_DEFAULT)?,
+            write_iter: gen_iter(CF_WRITE)?,
+            observer: ob,
+        })
+    }
+
+    fn next_item(&mut self) -> Result<Option<Ob::Target>> {
+        let mut lock_ok = box_try!(self.lock_iter.valid());
+        let mut writes_ok = box_try!(self.write_iter.valid());
+
+        let prefix = match (lock_ok, writes_ok) {
+            (false, false) => return Ok(None),
+            (true, false) => self.lock_iter.key(),
+            (false, true) => box_try!(Key::truncate_ts_for(self.write_iter.key())),
+            (true, true) => {
+                let prefix1 = self.lock_iter.key();
+                let prefix2 = box_try!(Key::truncate_ts_for(self.write_iter.key()));
+                match prefix1.cmp(prefix2) {
+                    Ordering::Less => {
+                        writes_ok = false;
+                        prefix1
+                    }
+                    Ordering::Greater => {
+                        lock_ok = false;
+                        prefix2
+                    }
+                    Ordering::Equal => prefix1,
+                }
+            }
+        };
+        self.observer.on_new_item(prefix);
+
+        while writes_ok {
+            let (key, value) = (self.write_iter.key(), self.write_iter.value());
+            writes_ok = self.observer.on_write(key, value)? && box_try!(self.write_iter.next());
+        }
+        while lock_ok {
+            let (key, value) = (self.lock_iter.key(), self.lock_iter.value());
+            lock_ok = self.observer.on_lock(key, value)? && box_try!(self.lock_iter.next());
+        }
+
+        let mut ok = box_try!(self.default_iter.valid());
+        while ok {
+            let (key, value) = (self.default_iter.key(), self.default_iter.value());
+            ok = self.observer.on_default(key, value)? && box_try!(self.default_iter.next());
+        }
+
+        Ok(Some(self.observer.emit()))
+    }
+}
+
+#[derive(Clone, Default)]
+struct MvccInfoCollector {
```

Comment:
Any test for `MvccInfoCollector`?

Reply:
Yes, tests are in `src/server/debug.rs`. The code is moved from that file.